storm-core/src/clj/backtype/storm/daemon/worker.clj

```diff
@@ -139,12 +139,12 @@
                   (.add local pair) 
 
                   ;;Using java objects directly to avoid performance issues in java code
-                  (let [node+port (get @task->node+port task)]
-                    (when (not (.get remoteMap node+port))
-                      (.put remoteMap node+port (ArrayList.)))
-                    (let [remote (.get remoteMap node+port)]
+                  (let []
+                    (when (not (.get remoteMap task))
+                      (.put remoteMap task (ArrayList.)))
+                    (let [remote (.get remoteMap task)]
                       (.add remote (TaskMessage. task (.serialize serializer tuple)))
-                     )))) 
+                    ))))
```

Comment:
@d2r 
Accessing task->node+port from these lines is root of this problem since it can be changed during preparing for transfer. So we should change current implementation to not accessing task->node+port.
It should be grouped by destination (node and port) within read lock.